File: src/brs/at/AtApiPlatformImpl.java
package brs.at;

import brs.Appendix;
import brs.Asset;
import brs.Attachment;
import brs.Attachment.ColoredCoinsAssetTransfer;
import brs.Burst;
import brs.Constants;
import brs.Transaction;
import brs.TransactionType;
import brs.crypto.Crypto;
import brs.fluxcapacitor.FluxValues;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.security.MessageDigest;
import java.util.Arrays;

public class AtApiPlatformImpl extends AtApiImpl {

    private static final Logger logger = LoggerFactory.getLogger(AtApiPlatformImpl.class);

    private static final AtApiPlatformImpl instance = new AtApiPlatformImpl();


    private AtApiPlatformImpl() {
    }

    public static AtApiPlatformImpl getInstance() {
        return instance;
    }

    private static Long findTransaction(int startHeight, int endHeight, Long atID, int numOfTx, long minAmount) {
        return Burst.getStores().getAtStore().findTransaction(startHeight, endHeight, atID, numOfTx, minAmount);
    }

    private static int findTransactionHeight(Long transactionId, int height, Long atID, long minAmount) {
        return Burst.getStores().getAtStore().findTransactionHeight(transactionId, height, atID, minAmount);
    }

    @Override
    public long getBlockTimestamp(AtMachineState state) {
        int height = state.getHeight();
        return AtApiHelper.getLongTimestamp(height, 0);
    }

    @Override
    public long getCreationTimestamp(AtMachineState state) {
        return AtApiHelper.getLongTimestamp(state.getCreationBlockHeight(), 0);
    }

    @Override
    public long getLastBlockTimestamp(AtMachineState state) {
        int height = state.getHeight() - 1;
        return AtApiHelper.getLongTimestamp(height, 0);
    }

    @Override
    public void putLastBlockHashInA(AtMachineState state) {
        ByteBuffer b = ByteBuffer.allocate(state.getA1().length * 4);
        b.order(ByteOrder.LITTLE_ENDIAN);

        b.put(Burst.getBlockchain().getBlockAtHeight(state.getHeight() - 1).getBlockHash());

        b.clear();

        byte[] temp = new byte[8];

        b.get(temp, 0, 8);
        state.setA1(temp);

        b.get(temp, 0, 8);
        state.setA2(temp);

        b.get(temp, 0, 8);
        state.setA3(temp);

        b.get(temp, 0, 8);
        state.setA4(temp);
    }

    @Override
    public void aToTxAfterTimestamp(long val, AtMachineState state) {

        int height = AtApiHelper.longToHeight(val);
        int numOfTx = AtApiHelper.longToNumOfTx(val);

        byte[] b = state.getId();

        long tx = findTransaction(height, state.getHeight(), AtApiHelper.getLong(b), numOfTx, state.minActivationAmount());
        logger.debug("tx with id {} found", tx);
        clearA(state);
        state.setA1(AtApiHelper.getByteArray(tx));

    }

    @Override
    public long getTypeForTxInA(AtMachineState state) {
        long txid = AtApiHelper.getLong(state.getA1());

        Transaction tx = Burst.getBlockchain().getTransaction(txid);

        if (tx == null || (tx.getHeight() >= state.getHeight())) {
            return -1;
        }

        if (tx.getMessage() != null) {
            return 1;
        }

        if (state.getVersion() >= 3) {
          return tx.getType().getType();
        }

        return 0;
    }

    @Override
    public long getAmountForTxInA(AtMachineState state) {
        long txId = AtApiHelper.getLong(state.getA1());

        Transaction tx = Burst.getBlockchain().getTransaction(txId);

        if (tx == null || (tx.getHeight() >= state.getHeight())) {
            return -1;
        }

        if (state.getVersion() >= 3) {
          long assetId = AtApiHelper.getLong(state.getA2());
          if (assetId != 0 && tx.getAttachment() instanceof Attachment.ColoredCoinsAssetTransfer) {
            Attachment.ColoredCoinsAssetTransfer assetTransfer = (ColoredCoinsAssetTransfer) tx.getAttachment();
            if(assetTransfer.getAssetId() == assetId) {
              return assetTransfer.getQuantityQNT();
            }
          }
        }
        if ((tx.getMessage() == null || Burst.getFluxCapacitor().getValue(FluxValues.AT_FIX_BLOCK_2, state.getHeight())) && state.minActivationAmount() <= tx.getAmountNQT()) {
            return tx.getAmountNQT() - state.minActivationAmount();
        }

        return 0;
    }

    @Override
    public long getMapValueKeysInA(AtMachineState state) {
      long key1 = AtApiHelper.getLong(state.getA1());
      long key2 = AtApiHelper.getLong(state.getA2());

      long atId = AtApiHelper.getLong(state.getA3());
      if(atId == 0L) {
        atId = AtApiHelper.getLong(state.getId());
      }

      return Burst.getStores().getAtStore().getMapValue(atId, key1, key2);
    }

    @Override
    public void setMapValueKeysInA(AtMachineState state) {
      long key1 = AtApiHelper.getLong(state.getA1());
      long key2 = AtApiHelper.getLong(state.getA2());
      long value = AtApiHelper.getLong(state.getA4());

      long atId = AtApiHelper.getLong(state.getId());

      state.addMapUpdate(new AT.AtMapEntry(atId, key1, key2, value));
    }

    @Override
    public long getTimestampForTxInA(AtMachineState state) {
        long txId = AtApiHelper.getLong(state.getA1());
        logger.debug("get timestamp for tx with id {} found", txId);
        Transaction tx = Burst.getBlockchain().getTransaction(txId);

        if (tx == null || (tx.getHeight() >= state.getHeight())) {
            return -1;
        }

        byte[] b = state.getId();
        int blockHeight = tx.getHeight();
        int txHeight = findTransactionHeight(txId, blockHeight, AtApiHelper.getLong(b), state.minActivationAmount());

        return AtApiHelper.getLongTimestamp(blockHeight, txHeight);
    }

    @Override
    public long getRandomIdForTxInA(AtMachineState state) {
        long txId = AtApiHelper.getLong(state.getA1());

        Transaction tx = Burst.getBlockchain().getTransaction(txId);

        if (tx == null || (tx.getHeight() >= state.getHeight())) {
            return -1;
        }

        int txBlockHeight = tx.getHeight();
        int blockHeight = state.getHeight();

        if (blockHeight - txBlockHeight < AtConstants.getInstance().blocksForRandom(blockHeight)) { //for tests - for real case 1440
            state.setWaitForNumberOfBlocks((int) AtConstants.getInstance().blocksForRandom(blockHeight) - (blockHeight - txBlockHeight));
            state.getMachineState().pc -= 7;
            state.getMachineState().stopped = true;
            return 0;
        }

        MessageDigest digest = Crypto.sha256();

        byte[] senderPublicKey = tx.getSenderPublicKey();

        ByteBuffer bf = ByteBuffer.allocate((Burst.getFluxCapacitor().getValue(FluxValues.SODIUM)) ?
        		32 + 8 + senderPublicKey.length :
        		32 + Long.SIZE + senderPublicKey.length);
        bf.order(ByteOrder.LITTLE_ENDIAN);
        bf.put(Burst.getBlockchain().getBlockAtHeight(blockHeight - 1).getGenerationSignature());
        bf.putLong(tx.getId());
        bf.put(senderPublicKey);

        digest.update(bf.array());
        byte[] byteRandom = digest.digest();

        return Math.abs(AtApiHelper.getLong(Arrays.copyOfRange(byteRandom, 0, 8)));
    }

    @Override
    public void messageFromTxInAToB(AtMachineState state) {
        long txid = AtApiHelper.getLong(state.getA1());

        Transaction tx = Burst.getBlockchain().getTransaction(txid);
        if (tx != null && tx.getHeight() >= state.getHeight()) {
            tx = null;
        }

        ByteBuffer b = ByteBuffer.allocate(state.getA1().length * 4);
        b.order(ByteOrder.LITTLE_ENDIAN);
        if (tx != null) {
            Appendix.Message txMessage = tx.getMessage();
            if (txMessage != null) {
                byte[] message = txMessage.getMessageBytes();
                if (message.length <= state.getA1().length * 4) {
                    b.put(message);
                }
            }
        }

        b.clear();

        byte[] temp = new byte[8];

        b.get(temp, 0, 8);
        state.setB1(temp);

        b.get(temp, 0, 8);
        state.setB2(temp);

        b.get(temp, 0, 8);
        state.setB3(temp);

        b.get(temp, 0, 8);
        state.setB4(temp);

    }

    @Override
    public void bToAddressOfTxInA(AtMachineState state) {
        long txId = AtApiHelper.getLong(state.getA1());

        clearB(state);

        Transaction tx = Burst.getBlockchain().getTransaction(txId);
        if (tx != null && tx.getHeight() >= state.getHeight()) {
            tx = null;
        }
        if (tx != null) {
            long address = tx.getSenderId();
            state.setB1(AtApiHelper.getByteArray(address));
        }
    }

    @Override
    public void bToAddressOfCreator(AtMachineState state) {
        long creator = AtApiHelper.getLong(state.getCreator());

        clearB(state);

        state.setB1(AtApiHelper.getByteArray(creator));

    }

    @Override
    public void putLastBlockGenerationSignatureInA(AtMachineState state) {
        ByteBuffer b = ByteBuffer.allocate(state.getA1().length * 4);
        b.order(ByteOrder.LITTLE_ENDIAN);

        b.put(Burst.getBlockchain().getBlockAtHeight(state.getHeight() - 1).getGenerationSignature());

        b.clear();

        byte[] temp = new byte[8];

        b.get(temp, 0, 8);
        state.setA1(temp);

        b.get(temp, 0, 8);
        state.setA2(temp);

        b.get(temp, 0, 8);
        state.setA3(temp);

        b.get(temp, 0, 8);
        state.setA4(temp);
    }

    @Override
    public long getCurrentBalance(AtMachineState state) {
        if (!Burst.getFluxCapacitor().getValue(FluxValues.AT_FIX_BLOCK_2, state.getHeight())) {
            return 0;
        }

        if (state.getVersion() >= 3) {
          long assetId = AtApiHelper.getLong(state.getB2());
          return state.getgBalance(assetId);
        }

        return state.getgBalance();
    }

    @Override
    public long getPreviousBalance(AtMachineState state) {
        if (!Burst.getFluxCapacitor().getValue(FluxValues.AT_FIX_BLOCK_2, state.getHeight())) {
            return 0;
        }

        return state.getpBalance();
    }

    @Override
    public void sendToAddressInB(long val, AtMachineState state) {
        if (val < 1)
            return;

        if (state.getVersion() > 2) {
          long assetId = AtApiHelper.getLong(state.getB2());
          long balance = state.getgBalance(assetId);
          if (val > balance) {
            val = balance;
          }
          AtTransaction tx = new AtTransaction(TransactionType.ColoredCoins.ASSET_TRANSFER,
              state.getId(), state.getB1().clone(), val, assetId, null);
          state.addTransaction(tx);

          state.setgBalance(assetId, balance - val);
        }

        if (val < state.getgBalance()) {
            AtTransaction tx = new AtTransaction(TransactionType.Payment.ORDINARY, state.getId(), state.getB1().clone(), val, 0L, null);
            state.addTransaction(tx);

            state.setgBalance(state.getgBalance() - val);
        } else {
            AtTransaction tx = new AtTransaction(TransactionType.Payment.ORDINARY, state.getId(), state.getB1().clone(), state.getgBalance(), 0L, null);
            state.addTransaction(tx);

            state.setgBalance(0L);
        }
    }

    @Override
    public void mintAsset(AtMachineState state) {
      long assetId = AtApiHelper.getLong(state.getB2());
      long accountId = AtApiHelper.getLong(state.getId());
      long quantity = AtApiHelper.getLong(state.getB1());

      Asset asset = Burst.getStores().getAssetStore().getAsset(assetId);
      if (asset == null || asset.getAccountId() != accountId) {
        // only assets that we have created internally
        return;
      }

      long circulatingSupply = Burst.getAssetExchange().getAssetCirculatingSupply(asset, false);
      long newSupply = circulatingSupply + quantity;
      if (newSupply > Constants.MAX_ASSET_QUANTITY_QNT) {
        // do not mint extra to keep the limit
        return;
      }

      AtTransaction tx = new AtTransaction(TransactionType.ColoredCoins.ASSET_MINT,
          state.getId(), state.getB1().clone(), quantity, assetId, null);
      state.addTransaction(tx);

      state.setgBalance(assetId, state.getgBalance(assetId) + quantity);
    }

    @Override
    public long issueAsset(AtMachineState state) {
      ByteBuffer b = ByteBuffer.allocate(32);
      b.order(ByteOrder.LITTLE_ENDIAN);
      b.put(state.getA1());
      b.put(state.getA2());
      b.put(state.getA3());
      b.put(state.getA4());
      b.clear();
      
      long decimals = AtApiHelper.getLong(state.getB1());

      AtTransaction tx = new AtTransaction(TransactionType.ColoredCoins.ASSET_ISSUANCE, state.getId(), null, decimals, 0L, b.array());
      state.addTransaction(tx);

      return tx.getAssetId();
    }

    @Override
    public void sendAllToAddressInB(AtMachineState state) {
        AtTransaction tx = new AtTransaction(TransactionType.Payment.ORDINARY, state.getId(), state.getB1().clone(), state.getgBalance(), 0L, null);
        state.addTransaction(tx);
        state.setgBalance(0L);
    }

    @Override
    public void sendOldToAddressInB(AtMachineState state) {
        if (state.getpBalance() > state.getgBalance()) {
            AtTransaction tx = new AtTransaction(TransactionType.Payment.ORDINARY, state.getId(), state.getB1(), state.getgBalance(), 0L, null);
            state.addTransaction(tx);

            state.setgBalance(0L);
            state.setpBalance(0L);
        } else {
            AtTransaction tx = new AtTransaction(TransactionType.Payment.ORDINARY, state.getId(), state.getB1(), state.getpBalance(), 0L, null);
            state.addTransaction(tx);

            state.setgBalance(state.getgBalance() - state.getpBalance());
            state.setpBalance(0L);
        }
    }

    @Override
    public void sendAToAddressInB(AtMachineState state) {
        ByteBuffer b = ByteBuffer.allocate(32);
        b.order(ByteOrder.LITTLE_ENDIAN);
        b.put(state.getA1());
        b.put(state.getA2());
        b.put(state.getA3());
        b.put(state.getA4());
        b.clear();

        AtTransaction tx = new AtTransaction(TransactionType.Payment.ORDINARY, state.getId(), state.getB1(), 0L, 0L, b.array());
        state.addTransaction(tx);
    }

    @Override
    public long addMinutesToTimestamp(long val1, long val2, AtMachineState state) {
        int height = AtApiHelper.longToHeight(val1);
        int numOfTx = AtApiHelper.longToNumOfTx(val1);
        int addHeight = height + (int) (val2 / AtConstants.getInstance().averageBlockMinutes(state.getHeight()));

        return AtApiHelper.getLongTimestamp(addHeight, numOfTx);
    }
}

File: src/main/kotlin/brs/at/AtApiPlatformImpl.kt
package brs.at

import brs.at.AtApi.Companion.REGISTER_PART_SIZE
import brs.at.AtApi.Companion.REGISTER_SIZE
import brs.entity.DependencyProvider
import brs.objects.FluxValues
import brs.util.byteArray.zero
import brs.util.convert.toUnsignedString
import brs.util.crypto.Crypto
import brs.util.logging.safeDebug
import org.slf4j.LoggerFactory
import java.nio.BufferUnderflowException
import kotlin.math.abs

class AtApiPlatformImpl constructor(private val dp: DependencyProvider) : AtApiImpl(dp) {
    override fun getBlockTimestamp(state: AtMachineState): Long {
        val height = state.height
        return AtApiHelper.getLongTimestamp(height, 0)
    }

    override fun getCreationTimestamp(state: AtMachineState): Long {
        return AtApiHelper.getLongTimestamp(state.creationBlockHeight, 0)
    }

    override fun getLastBlockTimestamp(state: AtMachineState): Long {
        val height = state.height - 1
        return AtApiHelper.getLongTimestamp(height, 0)
    }

    override fun putLastBlockHashInA(state: AtMachineState) {
        state.putInA(dp.blockchainService.getBlockAtHeight(state.height - 1)!!.hash)
    }

    override fun aToTxAfterTimestamp(value: Long, state: AtMachineState) {
        val height = AtApiHelper.longToHeight(value)
        val numOfTx = AtApiHelper.longToNumOfTx(value)

        val b = state.id

        val tx = dp.atStore.findTransaction(
            height,
            state.height,
            AtApiHelper.getLong(b!!),
            numOfTx,
            state.minActivationAmount()
        )!!
        logger.safeDebug { "tx with id ${tx.toUnsignedString()} found" }
        clearA(state)
        AtApiHelper.getByteArray(tx, state.a1)
    }

    override fun getTypeForTxInA(state: AtMachineState): Long {
        val txid = AtApiHelper.getLong(state.a1)

        val tx = dp.blockchainService.getTransaction(txid)

        if (tx == null || tx.height >= state.height) {
            return -1
        }

        return if (tx.message != null) {
            1
        } else 0

    }

    override fun getAmountForTxInA(state: AtMachineState): Long {
        val txId = AtApiHelper.getLong(state.a1)

        val tx = dp.blockchainService.getTransaction(txId)

        if (tx == null || tx.height >= state.height) {
            return -1
        }

        return if ((tx.message == null || dp.fluxCapacitorService.getValue(
                FluxValues.AT_FIX_BLOCK_2,
                state.height
            )) && state.minActivationAmount() <= tx.amountPlanck
        ) {
            tx.amountPlanck - state.minActivationAmount()
        } else 0

    }

    override fun getTimestampForTxInA(state: AtMachineState): Long {
        val txId = AtApiHelper.getLong(state.a1)
        logger.safeDebug { "get timestamp for tx with id ${txId.toUnsignedString()} found" }
        val tx = dp.blockchainService.getTransaction(txId)

        if (tx == null || tx.height >= state.height) {
            return -1
        }

        val b = state.id
        val blockHeight = tx.height
        val txHeight =
            dp.atStore.findTransactionHeight(txId, blockHeight, AtApiHelper.getLong(b!!), state.minActivationAmount())

        return AtApiHelper.getLongTimestamp(blockHeight, txHeight)
    }

    override fun getRandomIdForTxInA(state: AtMachineState): Long {
        val txId = AtApiHelper.getLong(state.a1)

        val tx = dp.blockchainService.getTransaction(txId)

        if (tx == null || tx.height >= state.height) {
            return -1
        }

        val txBlockHeight = tx.height
        val blockHeight = state.height

        if (blockHeight - txBlockHeight < dp.atConstants.blocksForRandom(blockHeight)) { //for tests - for real case 1440
            state.waitForNumberOfBlocks =
                dp.atConstants.blocksForRandom(blockHeight).toInt() - (blockHeight - txBlockHeight)
            state.machineState.pc -= 7
            state.machineState.stopped = true
            return 0
        }

        val digest = Crypto.sha256()
        digest.update(dp.blockchainService.getBlockAtHeight(blockHeight - 1)!!.generationSignature)
        digest.update(AtApiHelper.getByteArray(tx.id))
        digest.update(tx.senderPublicKey)
        digest.update(ByteArray(56)) // FIXME remove this in next fork
        val byteRandom = digest.digest()

        // TODO don't copy of range
        return abs(AtApiHelper.getLong(byteRandom.copyOfRange(0, REGISTER_PART_SIZE)))
    }

    override fun messageFromTxInAToB(state: AtMachineState) {
        val txid = AtApiHelper.getLong(state.a1)

        var tx = dp.blockchainService.getTransaction(txid)
        if (tx != null && tx.height >= state.height) {
            tx = null
        }

        if (tx != null) {
            val txMessage = tx.message
            if (txMessage != null) {
                var message = txMessage.messageBytes
                if (message.size <= REGISTER_SIZE) {
                    if (message.size < REGISTER_SIZE) {
                        val newMessage = ByteArray(REGISTER_SIZE)
                        message.copyInto(newMessage)
                        message = newMessage
                    }
                    state.putInB(message)
                    return
                }
            }
        }
        state.b1.zero()
        state.b2.zero()
        state.b3.zero()
        state.b4.zero()
    }

    override fun bToAddressOfTxInA(state: AtMachineState) {
        val txId = AtApiHelper.getLong(state.a1)

        clearB(state)

        var tx = dp.blockchainService.getTransaction(txId)
        if (tx != null && tx.height >= state.height) {
            tx = null
        }
        if (tx != null) {
            val address = tx.senderId
            AtApiHelper.getByteArray(address, state.b1)
        }
    }

    override fun bToAddressOfCreator(state: AtMachineState) {
        val creator = AtApiHelper.getLong(state.creator!!)

        clearB(state)

        AtApiHelper.getByteArray(creator, state.b1)
    }

    override fun putLastBlockGenerationSignatureInA(state: AtMachineState) {
        if (dp.fluxCapacitorService.getValue(FluxValues.NEXT_FORK)) {
            state.putInA(dp.blockchainService.getBlockAtHeight(state.height - 1)!!.generationSignature)
        } else {
            // Fast fail.
            throw BufferUnderflowException()
        }
    }

    override fun getCurrentBalance(state: AtMachineState): Long {
        return if (!dp.fluxCapacitorService.getValue(FluxValues.AT_FIX_BLOCK_2, state.height)) {
            0
        } else state.getgBalance()

    }

    override fun getPreviousBalance(state: AtMachineState): Long {
        return if (!dp.fluxCapacitorService.getValue(FluxValues.AT_FIX_BLOCK_2, state.height)) {
            0
        } else state.getpBalance()
    }

    override fun sendToAddressInB(value: Long, state: AtMachineState) {
        if (value < 1)
            return

        if (value < state.getgBalance()) {
            val tx = AtTransaction(state.id!!, state.b1.clone(), value, null)
            state.addTransaction(tx)

            state.setgBalance(state.getgBalance() - value)
        } else {
            val tx = AtTransaction(state.id!!, state.b1.clone(), state.getgBalance(), null)
            state.addTransaction(tx)

            state.setgBalance(0L)
        }
    }

    override fun sendAllToAddressInB(state: AtMachineState) {
        val tx = AtTransaction(state.id!!, state.b1.clone(), state.getgBalance(), null)
        state.addTransaction(tx)
        state.setgBalance(0L)
    }

    override fun sendOldToAddressInB(state: AtMachineState) {
        if (state.getpBalance() > state.getgBalance()) {
            val tx = AtTransaction(state.id!!, state.b1, state.getgBalance(), null)
            state.addTransaction(tx)

            state.setgBalance(0L)
            state.setpBalance(0L)
        } else {
            val tx = AtTransaction(state.id!!, state.b1, state.getpBalance(), null)
            state.addTransaction(tx)

            state.setgBalance(state.getgBalance() - state.getpBalance())
            state.setpBalance(0L)
        }
    }

    override fun sendAToAddressInB(state: AtMachineState) {
        val b = ByteArray(REGISTER_SIZE)
        state.a1.copyInto(b, 0)
        state.a2.copyInto(b, 8)
        state.a3.copyInto(b, 16)
        state.a4.copyInto(b, 24)

        val tx = AtTransaction(state.id!!, state.b1, 0L, b)
        state.addTransaction(tx)
    }

    override fun addMinutesToTimestamp(val1: Long, val2: Long, state: AtMachineState): Long {
        val height = AtApiHelper.longToHeight(val1)
        val numOfTx = AtApiHelper.longToNumOfTx(val1)
        val addHeight = height + (val2 / dp.atConstants.averageBlockMinutes(state.height)).toInt()

        return AtApiHelper.getLongTimestamp(addHeight, numOfTx)
    }

    companion object {
        private val logger = LoggerFactory.getLogger(AtApiPlatformImpl::class.java)
    }
}
